File: src/services/content_safety.py
from azure.ai.contentsafety import ContentSafetyClient
from azure.core.credentials import AzureKeyCredential
from azure.core.exceptions import HttpResponseError
from azure.ai.contentsafety.models import AnalyzeTextOptions, TextCategory, AnalyzeImageOptions, ImageData

def content_analyze(image, text):

    # Create a Content Safety client
    client = ContentSafetyClient(CONFIG["content_moderator"]["endpoint"], AzureKeyCredential(CONFIG["content_moderator"]["key"]))

    try:
        result_image = None;
        result_text = None;    

        if image: 
           result_image = is_inapproprieate_content(client.analyze_image(AnalyzeImageOptions(image=ImageData(content=image))))

        if text:
            result_text = is_inapproprieate_content(client.analyze_text( AnalyzeTextOptions(text=text))) 

        return result_image or result_text

    except Exception as ex:
        print(f'Error when checking inappropriate content in text/image  -> {ex}')

def is_inapproprieate_content(response):
    return (next(item for item in response.categories_analysis if item.category == TextCategory.HATE).severity > 0) or \
           (next(item for item in response.categories_analysis if item.category == TextCategory.SELF_HARM).severity > 0) or \
           (next(item for item in response.categories_analysis if item.category == TextCategory.SEXUAL).severity > 0) or \
           (next(item for item in response.categories_analysis if item.category == TextCategory.VIOLENCE).severity > 0)

File: Services/ContentSafety.cs
using Azure;
using Azure.AI.ContentSafety;
using Microsoft.AspNetCore.DataProtection.KeyManagement;

namespace hackaton_microsoft_agro.Services
{
    public class ContentSafety(string endpoint, string key)
    {
        ContentSafetyClient contentSafetyClient = new ContentSafetyClient(new Uri(endpoint), new AzureKeyCredential(key));
        BlocklistClient blocklistClient = new BlocklistClient(new Uri(endpoint), new AzureKeyCredential(key));

        public bool ContentAnalyze(byte[] image, string text)
        {
            bool resultImage = false;
            bool resultText = false;

            try
            {
                if (image != null)
                {
                    var request = new AnalyzeImageOptions(new ContentSafetyImageData(BinaryData.FromBytes(image)));
                    resultImage = IsInappropriateImage(contentSafetyClient.AnalyzeImage(request));
                }

                if (text != null)
                {
                    var request = new AnalyzeTextOptions(text);
                    resultText = IsInappropriateText(contentSafetyClient.AnalyzeText(request));
                }
            }
            catch (Exception ex)
            {
                throw new Exception("Error ContentAnalyze -> " + ex.Message);
            }

            return resultImage || resultText;
        }

        bool IsInappropriateImage(Response<AnalyzeImageResult> response)
        {
            return (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == ImageCategory.Hate)?.Severity > 0) ||
            (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == ImageCategory.SelfHarm)?.Severity > 0) ||
            (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == ImageCategory.Sexual)?.Severity > 0) ||
            (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == ImageCategory.Violence)?.Severity > 0);
        }

        bool IsInappropriateText(Response<AnalyzeTextResult> response)
        {
            return (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == TextCategory.Hate)?.Severity > 0) ||
            (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == TextCategory.SelfHarm)?.Severity > 0) ||
            (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == TextCategory.Sexual)?.Severity > 0) ||
            (response.Value.CategoriesAnalysis.FirstOrDefault(item => item.Category == TextCategory.Violence)?.Severity > 0);
        }
    }
}
